### ISSUE TITLE
Make proxy code reusable

### ISSUE THREAD

**Issue by CodeZombieCH** [Microsoft/vscode#12588]
Proxy support for Visual Studio Code has been introduced recently through environment variables/settings (see https://code.visualstudio.com/docs/setup/setup-overview#_proxy-server-support). As an extension developer I would like to use these settings when requesting resources from the internet. My current implementation fails for users using a proxy.

What is the recommended way for an extension to use a proxy?

I could not find a `vscode` API that exposes methods to make requests using a proxy other than access to the proxy settings (`workspace.getConfiguration()`) and implementing everything on my own. Looking at the Visual Studio Code code base I identified the code that adds support for proxies, but copy&paste it into my extension would result in duplicated and potentially outdated code, which is something I want to avoid whenever possible. As I am probably not the only extension developer accessing the internet, I would like to hear your opinion on this issue. Any feedback would be greatly appreciated.

**Comment by jrieken** [comment]
-1 for adding this for to the API but +1 for extracting our code into a node-module. @joaomoreno how feasible is that? how much code do we add on top of other node-modules?

**Comment by joaomoreno** [comment]
It's thin.

**Comment by jrieken** [comment]
I leaving it with you then ;-)

**Comment by CodeZombieCH** [comment]
A node package sounds like a cool solution. Let me know when it's ready so I can test it and provide feedback.

**Comment by joaomoreno** [comment]
We're going with a different proxy solution in the future, skipping this.

**Comment by CodeZombieCH** [comment]
Do you have any details about that future solution?

**Comment by joaomoreno** [comment]
Actually, I might've closed this too eagerly.

**Comment by tebeco** [comment]
@joaomoreno any news from April ?
times to times, We have to set our password in clear so far in the settings json file in order to update OmniSharp

This is insane and quite critical
Everybody tries to point someone else team and it really hurt lots of people using VsCode in companies

**Comment by tebeco** [comment]
Do i need to ping more contributor to VsCode in order to get more feedback on it ?

**Comment by Stibbons** [comment]
+1

**Comment by genenorton** [comment]
This is super annoying...  It makes extensions mostly useless in corporate world.  which makes it unlikely that I'll use vs code  at work.....  back to notepad++ i guess.....

++1

**Comment by tebeco** [comment]
+1
VsCode need to proxy an API to extensions that provide a kind of “fetch” api but will read proxy settings if necessary especially when on windows this settings is ALREADY used by vscode itself

**Comment by tebeco** [comment]
As a reminder in corporate having to set the password in clear is not ok at all
same goes for HTTP_PROXY no go

and CNTLM is a crap on windows, ever tried npm install behind ? nop

**Comment by PlanetWilson** [comment]
This is getting really painful. As VSCode gains in popularity and is the de-facto standard for Office365 dev these days it is critical to get this working.

**Comment by vicpara** [comment]
Terminate Visual Studio Code , go in terminal and run `code  --ignore-certificate-errors` to start it ignoring ssl certificates. This solved my problem.

**Comment by jeannich** [comment]
The solution proposed by @vicpara above did not work for me.

Instead I am using this workaround:
`code --proxy-server="http=proxy:8080;https=proxy:8080;ftp=proxy:8080"`

**Comment by tebeco** [comment]
of course it does not work :D
ignoring ssl certificate is just if on top of having a proxy, your proxy actually rewrite ssl with it own certificate and that certificate is a self signed certificate 

vscode is still lacking the ability to check 

`hkcu:/software/microsoft/windows/currentversion/internet settings`

then goes for keys like
`AutoConfigUrl` for PAC script 
`ProxyServer` for http proxy

in fact all the data are here
`hkcu:/software/microsoft/windows/currentversion/internet settings/Connections/DefaultConnectionSettings`
need to check that bitarray with :
array[8] <== this is a bit mask and this tells you what to read (what is enabled or not)

**Comment by tebeco** [comment]
this is what user on windows expect in fact
you should not do trick running vsCode when the soft it self build target that windows could actually process these already existing values

**Comment by m451** [comment]
I agree with tebeco. The system default proxy should be used if nothing else was configured. Still it's probably a good thing to have overrides in the workspace settings so devs working on the same project from the same environment can simply share the correct config. 

Yet it seems like there are many things missing. E.g. "proxy" is not the only thing that needs to be implemented for proper proxy support. The exclusion / bypass list of the OS also has to be correctly handled and we need a override setting for that as well. 

I'm aware of the chromium switches but they do not what the other two options do: 
- workspace setting options: enable quick sharing of a workspace config
- system os defaults enable a great out-of-box experiance 

Also I found the chromium parameters not to work all the time. In my case --proxy-server="myproxy:port" is ignored. It works if I set it in the workspace settings but then the --proxy-bypass-list parameter is ignored.

**Comment by kaufmd** [comment]
I am running the latest VS Code version on macOS Sierra and I'm seeing a `ECONNREFUSED` when trying to fetch from "https://schemastore.azurewebsites.net/schemas/json/tsconfig.json" whenever I open my `tsconfig.json` even though the proxy settings are definitely correct. When I call the failing address in a Chrome instance it opens perfectly fine.
I was testing using all the methods there are:
- starting VS Code with parameters
- setting the `http.proxy`, etc. settings within VS Code
- setting environment variables HTTP_PROXY and HTTPS_PROXY

My findings: 
1. on macOS the parameters at startup are having absolutely no effect on the behavior. VS Code cannot even resolve the URL.

2. Setting `http.proxy` inside of VS Code or using the env variables has the same effect (even though I didn't test what takes precedence in the end)

3. The `ECONNREFUSED` seems to come from our company proxy server. I could rule out that it is a authentication problem by specifying my credentials (which can't be a permanent solution anyway as pointed out by @tebeco)

Considering that these issues have been going on for almost 2 years now (look at the ticket creation date!) it is definitely about time to solve them for good. @joaomoreno considering that this also happens with the TypeScript extension is very sad to see. Microsoft should have an eager interest in supporting corporate environments. The lack of updates and support from the VS Code team in this ticket - sorry to say - though unfortunately shows the opposite.

**Comment by joshsackos** [comment]
Cannot install Particle Workbench dependencies behind a proxy at work.  This is an old ticket, if Microsoft actually wants people to use VS Code, then they should fix this.  Companies should really look at an alternative to VS Code due to this limitation.  All corporate users behind a proxy will not be able to use the company's VS Code extensions that need to download external content, thus rendering the company's efforts useless, and then the customer will probably find an alternative product that provides decent tools, which will reduce the company's profits.  Have a merry Christmas everyone! :)

**Comment by gsemet** [comment]
If you define HTTP_PROXY and HTTPS_PROXY, extension can « see » the outer world. It I sad to see this support is not straight but as least it does the job. And you may have to hardcore your proxy password in an env car, which is pretty bad...

**Comment by joshsackos** [comment]
No, setting HTTP_PROXY and HTTPS_PROXY does not solve the issue.

**Comment by gsemet** [comment]
I do have them, with lowercase as well, in my use env var, and it works

**Comment by joshsackos** [comment]
Windows environment variables are not case sensitive, so you can only have one variable for each.  Glad that you have a solution that works for you.  Obviously it is not a universal solution though.  Best of luck to you, happy holidays!

**Comment by joaomoreno** [comment]
Hey guys you might wanna follow https://github.com/Microsoft/vscode/issues/60773

There's been some great exploration development by @chrmarti: https://github.com/Microsoft/vscode/issues/64202

**Comment by nschonni** [comment]
It looks like this was included with 1.31.
It doesn't support NO_PROXY, but I'm opening a separate issue for that

**Comment by SetTrend** [comment]
@m451: I suggest to utilize the system default proxy, too.

The organization I'm currently working for is using NTLM authentication with their proxy. This is causing HTTP 407 errors when using environment variables `HTTP_PROXY`/`HTTPS_PROXY`.

**Comment by solarjoe** [comment]
I found this issue https://github.com/Microsoft/vscode/issues/38888 

They mention that supplying the proxy parameters on the command line works, even if setting it in VS Code doesn't. And this works for me. Still not beautiful, as the password is clear text, but for the other options (environment variables, settings.json) this would also be the case.

So I added this to my desktop shortcut and the two popups are gone:

`"C:\Path\To\Code.exe" --proxy-server=http://user:password@123.123.123.123:8080`

**Comment by tebeco** [comment]
pretty sure the password in the clear is a joke for everyone here.
I would love VsCode to know if it's possible that VsCode offers an Api to use the same proxy that the updater use. 
as long as i know the updater has no issue. the main issue is that i'm not sure NodeJs can handle Ntlm, so it would need VsCode to spawn a process to proxify request properly
then the settings would just be "use system proxy" (like most browser today i believe), that would check for `HTTP(S)_PROXY / NO_PROXY` Pac script etc ... well standard system Proxy

as vscode builds are already Platform specific i guess platform specific feature is something already possible ?

**Comment by chrmarti** [comment]
For downloading extensions and VS Code updates we are using Chromium's cross-platform support for proxies. That looks up the proxy configuration in the OS.

For extensions (which run in a separate process that does not have Chromium's network library available) we are using an Electron API to look up the proxy configuration from the OS. That is cross-platform, but it does not support authentication.

**Comment by tebeco** [comment]
thx a lot for the insight on how it work and what's available with electron/chromium.
As a general pattern with chromium/electron we often see dedicated child process per tasks (Network/GPU/ ...)

Having one dedicated to forward request using Platform specific API without the limit of Electron API (as we seems to all agree that Electron API is too limited for that issue, and require more code to have both proxy working and security / user handling :
* no password in clear text
* use Os settings
* respect entreprise GPO (pac script)
* user does not have to change it password every time she/he change AD login

As explicit/auth proxy is a disaster in lots of company that should replaced by implicit proxy, lots of users are suffering.

I'm just looking for a secure way to avoid clear text pas in vscode

**Comment by meustrus** [comment]
Ever since proxy support for extensions was enabled, my extensions stopped being able to reach the internet. Every attempt results in a certificate authentication error.

I'm guessing that my problem is that the patch for `https` does not also pick up the OS-level certificate store. This appears to be what the setting `http.systemCertificates` is for. **Does that setting also patch `https` for extensions?**

For now, turning off the patch for extensions with `"http.proxySupport": "off"` fixes my problem because the network allows some requests through without using the PAC-file based proxy that uses the corporate root CA. That won't solve the problem for everyone, though. microsoft/vscode-azurefunctions#1429 may be caused by a transparent proxy that *does* require the corporate root CA.

**Comment by chrmarti** [comment]
> I'm guessing that my problem is that the patch for `https` does not also pick up the OS-level certificate store. This appears to be what the setting `http.systemCertificates` is for. **Does that setting also patch `https` for extensions?**

Yes, `http.systemCertificates` is specifically supporting the `https` module for extensions. Could you open a new issue, so we can look into why this doesn't work in your case?

**Comment by kinwolfqc** [comment]
Ugh, 3 years and not close to a solution.  I guess I should tell management to forget about using VS Code.  We're behind an authentication proxy and as soon as an extension tries to download something, like the language server, it fails, so we cannot use them.

**Comment by GigiusB** [comment]
This is incredibly frustrating that such an issue should prevent our enterprise to adopt it !
There is no way we can do without the corporate proxy

**Comment by orthoxerox** [comment]
For those visiting this from a search engine, to make extensions like C# work via an https proxy with authentication, right now you need the following:

1. You need your proxy with username and password in settings.json: "http.proxy": "https://user:pwd@host:port" (without it you will get error 407)
2. You need to install win-ca extension from ukoloff
3. You need "win-ca": "append" in settings.json (without it you will get "Error: self signed certificate in certificate chain")
4. Don't put anything else proxy or http-related in settings.json

I tested this both in 1.46.1 and 1.48.0 (with auth.js fix)

**Comment by tebeco** [comment]
This is why it would be more sustainable to have all of this baked within VsCode, and not at the full charge of extensions developper, and at the cost of consumer security.

Somehow, having an API in VsCode return an HttpClient already setup, and supporting POSSIBLE `NTLMv2` / `Kerberos` look like something acheivable in order to use `VsCode` safetly.

The equivalent in C# would look like:

```csharp
public HttpClient GetMeSomethingReady()
{
  var proxyHttpClientHandler = new HttpClientHandler()
  {
    Proxy = new WebProxy("http://FROM_THE_SETTING_NO_CREDENTIALS")
    {
      Credentials = CredentialCache.DefaultNetworkCredentials
    },
    SslProtocols = SslProtocols.Tls13 | SslProtocols.Tls12 | SslProtocols.Tls11 | SslProtocols.Tls,
    UseProxy = true
  };

  return new HttpClient(proxyHttpClientHandler);
```

then extensions developers would do something like
```cs
var httpClient = GetMeSomethingThatWorks();
httpClient.Get(url)
```

If they want to do everything from scratch ... why not, but it look like this should be 
* out of the box
* 0 extensions required
* no password nowhere, as the user is already logged, and already has `NTLMv2` / `Kerberos` ticket
* use the User Certificate Store on windows build (that's the standard)
* No env var as this would require special script to run VS not to polute shell session or user session

**Comment by chrmarti** [comment]
@orthoxerox Part of win-ca is already built-in, you shouldn't need to install it separately.

**Comment by orthoxerox** [comment]
@chrmarti I shouldn't, but I had to. `"http.systemCertificates": true` is the default setting, but it only allows VSCode to download extensions without complaining about self-signed certs, not the extensions to download their dependencies.

**Comment by southwood** [comment]
Here is a workaround that enables extensions to be used in a devcontainer while a mitm solution (such as zscaler) is deployed to your network.

In `devcontainer.json`:
```json
"remoteEnv": {
    "NODE_EXTRA_CA_CERTS": "<PROVIDE_YOUR_TRUSTED_ROOT_CA_CERT_HERE>.crt"
},
"postCreateCommand": "sed -i -e 's/this\\.strictSSL=/this\\.strictSSL=false\\&\\&/g' $(find ~ -name *HostAgent.js)",
```

I think root cause is that remoteExtensionHostAgent.js ignores all the existing settings.json files and overrides local container's npm/node configuration by default. Another way to achieve this is to configure `http.proxyStringSSL` in your local settings.json and then drop that into every subdirectory of the extension runtime `find ~ -type d | xargs -I {} cp .vscode/settings.json {}`. If anyone knows which specific folder's settings.json is used by remoteExtensionHostAgent.js, that would be a better option than using sed to edit the minified js.

**Comment by Szaboadrii22** [comment]
In VSCode Settings, you can set http.proxy to http://username:PASSWORD@your-proxy-ip:port

**Comment by Szaboadrii22** [comment]
Extension proxy support #12588

**Comment by joewood** [comment]
One alternative to the suggestion of creating a VSCode fetch API for extensions would be to use the VSCode host as a proxy. VS Code already supports PAC files and NTLM/GSS corporate proxy authentication. Using this networking stack it could expose a port to the extensions like CNTLM. This could be exposed in the form of http_proxy env variables (so no change would be required there). Network requests would then pass through VS Code. 

This feels like the simplest approach as it requires no change to extensions (continue to use http_proxy), the code for port forwarding is already in present.

**Comment by tebeco** [comment]
that's why multiple users in that thread are asking for since 2016/2017

it seems to be most logical one in order to have something uniform and also consistent with the OS

**Comment by LeYom** [comment]
Hi
I am too behind a corporate proxy with kerberos authentifcation, and I didn't understand why Kerberos authentification for the Proxy HTTP Adapter is not supported. 

The problem is that we can't provide login/password because credentials are automatically manage by the OS Kerberos authentification system.

I hope this feature to be added to the roadmap, since some critical extensions doesn't work well offline today (like IntelliSense which needs to download its model to be fine)

Regards

**Comment by penenkel** [comment]
> Hi I am too behind a corporate proxy with kerberos authentifcation, and I didn't understand why Kerberos authentification for the Proxy HTTP Adapter is not supported.
> 
> The problem is that we can't provide login/password because credentials are automatically manage by the OS Kerberos authentification system.
> 
> I hope this feature to be added to the roadmap, since some critical extensions doesn't work well offline today (like IntelliSense which needs to download its model to be fine)
> 
> Regards

@LeYom  Of topic for this issue but perhaps a solution for your problem: Have you tried using a second, local proxy that supports NTLM or Kerberos authentication to the corporate proxy? I'm currently using [px](https://github.com/genotrance/px) which works quite well. An other, if somewhat aged, candidate is `cntlm`.

**Comment by tebeco** [comment]
> > Hi I am too behind a corporate proxy with kerberos authentifcation, and I didn't understand why Kerberos authentification for the Proxy HTTP Adapter is not supported.
> > The problem is that we can't provide login/password because credentials are automatically manage by the OS Kerberos authentification system.
> > I hope this feature to be added to the roadmap, since some critical extensions doesn't work well offline today (like IntelliSense which needs to download its model to be fine)
> > Regards
> 
> @LeYom Of topic for this issue but perhaps a solution for your problem: Have you tried using a second, local proxy that supports NTLM or Kerberos authentication to the corporate proxy? I'm currently using [px](https://github.com/genotrance/px) which works quite well. An other, if somewhat aged, candidate is `cntlm`.

oO i can answer for that one
avoid `cntlm` for workflow like that
the moment you'll run `npm install` it will start puking on itself and slowly die as it does not manage to sustain that much request even when attempting to configure it for more "parallelism"
or also fun issue with things like linux path syntax on windows 🤦‍♂️
or the fact that it must be in a specific path and if you change it ... cf point above🤷‍♂️

i don't recall the issue we struggled with `px` but it wasn't glorious either even if still better than `cntlm`

**Comment by RockyMM** [comment]
I hope the number of upvotes will bring this feature request closer to the reality ;)

**Comment by mckennahampton** [comment]
This is also where I ended up while researching why the official SQL Server extension was failing to download the latest SQL Tools service layer package (error "[ERROR] Error: Request error: EPROTO").

I've been running into similar issues with other extensions that try to download additional payloads once they're installed. They all fail if I'm behind a proxy. I've blanket whitelisted the URL's (which are mostly to github.com) and also bypassed all SSL decryption, but there is still some SSL/TLS cert negotiation issue at hand. Unfortunately it's stopping us (and I'm sure many others) from trying out other things like Copilot for VScode since it's the same problem.

**Comment by merlinz01** [comment]
@chrmarti Can we have an update on where we stand on this? An issue eight years old and so many people waiting should be high priority.

**Comment by chrmarti** [comment]
@merlinz01 Work is under way to make Chromium's implementation of `fetch` available to extensions. This issue is unlikely to be closed soon though, as many extensions are using Node's `https` module for which we have only partial support for proxies. Only proxies without auth and with Kerberos auth are supported at the moment.

If you think your setup should already work with the `https` module, but doesn't, please install the Network Proxy Test extension (https://marketplace.visualstudio.com/items?itemName=chrmarti.network-proxy-test) and check the output of `F1` > `Network Proxy Test: Test Connection` in VS Code.

**Comment by bmingles** [comment]
> [@merlinz01](https://github.com/merlinz01) Work is under way to make Chromium's implementation of `fetch` available to extensions. This issue is unlikely to be closed soon though, as many extensions are using Node's `https` module for which we have only partial support for proxies. Only proxies without auth and with Kerberos auth are supported at the moment.
> 
> If you think your setup should already work with the `https` module, but doesn't, please install the Network Proxy Test extension (https://marketplace.visualstudio.com/items?itemName=chrmarti.network-proxy-test) and check the output of `F1` > `Network Proxy Test: Test Connection` in VS Code.

@chrmarti Is there a GH issue outlining the Chromium `fetch` support that can be watched?

**Comment by chrmarti** [comment]
@bmingles We discovered that Chromium's `fetch` (or what Electron surfaces) is not feature-identical with Node.js' `fetch` and so we have moved on to add proxy support to Node.js' `fetch`. We expect that to offer better compatibility for code written for Node.js' `fetch` and have enabled it by default in the user settings:

<img width="512" alt="Image" src="https://github.com/user-attachments/assets/8870d5f7-ee8c-4752-95e0-2ae1b859aeea" />

Our work for making Chromium/Electron's `fetch` available can also be enabled in the user settings if you want to give that a try:

<img width="490" alt="Image" src="https://github.com/user-attachments/assets/828652f4-7b90-4576-824f-2d9ccd2e77d0" />

**Comment by merlinz01** [comment]
I don't know for sure if it's related, but whenever I try to install an extension by clicking the Install button, I get this cryptic error:
```
2025-02-05 13:10:12.021 [error] Error: net::ERR_INVALID_ARGUMENT
    at SimpleURLLoaderWrapper.<anonymous> (node:electron/js2c/utility_init:2:10511)
    at SimpleURLLoaderWrapper.emit (node:events:518:28)
2025-02-05 13:10:12.056 [error] net::ERR_INVALID_ARGUMENT: Error: net::ERR_INVALID_ARGUMENT
    at SimpleURLLoaderWrapper.<anonymous> (node:electron/js2c/utility_init:2:10511)
    at SimpleURLLoaderWrapper.emit (node:events:518:28)
```

The results of the network proxy test:
```
VS Code 1.96.4 (cd4ee3b1c348a13bafd8f9ad8060705f6d4b9cba)
Network Proxy Test 0.0.13
linux 6.8.0-52-generic x64

Settings:
- http.proxy: http://user:pass@proxy:port
  - globalValue: http://user:pass@proxy:port
- http.proxyAuthorization: user:pass
  - globalValue: user:pass
- http.proxyStrictSSL: false
  - globalValue: false
- http.proxySupport: override
- http.systemCertificates: true
- http.experimental.systemCertificatesV2: true
  - globalValue: true
- http.electronFetch: true
  - globalValue: true
- http.fetchAdditionalSupport: true

Environment variables:
HTTPS_PROXY=http://user:pass@proxy:port
HTTP_PROXY=http://user:pass@proxy:port
NO_PROXY=
http_proxy=http://user:pass@proxy:port
https_proxy=http://user:pass@proxy:port

Sending GET request to https://github.com...
vscode-proxy-agent: PROXY user:pass@proxy:port
Received response:
- Status: 200 OK
Certificate chain:
- Subject: github.com
  Subject alt: DNS:github.com
  Validity: Feb  4 05:00:00 2025 GMT - Feb  7 05:00:00 2025 GMT
  Fingerprint: 1B:E0:89:4E:D1:78:00:82:7D:67:28:95:31:B7:DE:80:B0:9B:A3:E6
 (Proxy certificates)
Local root certificates:
 (Proxy root certificate)

Sending GET request to https://github.com using fetch from Electron...
Received response:
- Status: 200 

Sending GET request to https://github.com using fetch from Node.js...
Received response:
- Status: 200 OK
```

Most of the other parts of VSCode work fine with the proxy.

**Comment by bmingles** [comment]
> [@bmingles](https://github.com/bmingles) We discovered that Chromium's `fetch` (or what Electron surfaces) is not feature-identical with Node.js' `fetch` and so we have moved on to add proxy support to Node.js' `fetch`. We expect that to offer better compatibility for code written for Node.js' `fetch` and have enabled it by default in the user settings:

> Our work for making Chromium/Electron's `fetch` available can also be enabled in the user settings if you want to give that a try:

Thanks @chrmarti,
We use self-signed certs for some of our servers and are currently setting the `NODE_EXTRA_CA_CERTS` env variable in the process running VS Code to include our signing CA certs. There are no proxies involved here, it just seems to be what is required to get NodeJS to trust additional certs.

Any chance these changes will make it possible for extensions to use the OS trust store instead of requiring `NODE_EXTRA_CA_CERTS` to be set, or does this only apply to proxy scenarios?

**Comment by chrmarti** [comment]
@merlinz01 Could you try with Electron Fetch turned off:
<img width="490" alt="Image" src="https://github.com/user-attachments/assets/828652f4-7b90-4576-824f-2d9ccd2e77d0" />

@bmingles You can add the certificate to the OS trusted root certificates, these should all be picked up. (Restart VS Code after changing the OS trust store.)

**Comment by merlinz01** [comment]
@chrmarti I tried turning off the Electron Fetch and restarting VSCode, but it gives the same error.

**Comment by bmingles** [comment]
@chrmarti I still need to do some more testing, but looks like the `Http: Electron Fetch` works for our scenario.
1. It now allows http2 connections via fetch in a spot where we were having to implement a custom transport based on node:http2
2. the NODE_EXTRA_CA_CERTS seems to no longer be needed

Thanks!!

Out of curiosity, is `Http: Electron Fetch` documented somewhere that I missed, or has it just not made it to the docs?

**Comment by chrmarti** [comment]
@merlinz01 Could you open a new issue for us to investigate?

@bmingles It is not documented beyond the settings itself. We have paused our effort to make it ready to become the default because we have found a way to add proxy support to Node.js' `fetch` and will likely keep that as the default. That also supports OS certificates and you can pass it an Agent from the undici library to enable http2 if required.

**Comment by SyaifulnizamshamsudinMax123** [comment]
> Proxy support for Visual Studio Code has been introduced recently through environment variables/settings (see https://code.visualstudio.com/docs/setup/setup-overview#_proxy-server-support). As an extension developer I would like to use these settings when requesting resources from the internet. My current implementation fails for users using a proxy.
> 
> What is the recommended way for an extension to use a proxy?
> 
> I could not find a `vscode` API that exposes methods to make requests using a proxy other than access to the proxy settings (`workspace.getConfiguration()`) and implementing everything on my own. Looking at the Visual Studio Code code base I identified the code that adds support for proxies, but copy&paste it into my extension would result in duplicated and potentially outdated code, which is something I want to avoid whenever possible. As I am probably not the only extension developer accessing the internet, I would like to hear your opinion on this issue. Any feedback would be greatly appreciated.
Thanks you.

**Comment by chrmarti** [comment]
@SyaifulnizamshamsudinMax123 Extensions automatically get the proxy support we added. Currently `fetch`, `http` and `https` all have proxy support added and most NPM packages for network requests use one of these. Please open a bug report if that doesn't work for your extension. Thanks.

**Comment by tiennou** [comment]
I'm getting a
```
Unable to load schema from 'http://json.schemastore.org/package': Not Found. The requested location could not be found.
Unable to connect to http://json.schemastore.org/package. Error: connect EBADF 20.42.128.105:80.
```
in any package.json file I open, and I'm not sure what's going on. I do not have a proxy (this is a residential network). Loading the URL through my browser correctly nets me what looks like a 40Kb schema file.
